### DISCUSSION TITLE
RTPSRelay Vertical Handler participant discovery

### DISCUSSION THREAD

**Discussion by smnrgrs** [OpenDDS/OpenDDS#4386] [Q&A] [answered]
Is it possible for the RTPSRelay to discover local RTPS participants using the standard mechanisms (RTPS Discovery/RTPS-UDP), as well as allowing the normal Vertical Handler TCP connections for remote participants behind NAT routers etc?
I have an RTPSRelay in the cloud which will correctly relay data to/from participants connected on its TCP connections, but won't relay data relating to local DDS participants running on the same VM. (I don't want to get these local participants to connect using SpdpRtpsRelayAddress lines).

**Comment by jrw972** [comment]
No.  An RtpsRelay instance consists of three components: the Handler, the Application Participant, and the Relay Participant.  The vertical ports belong to the Handler, and it performs the task of routing messages to connected client participants.  One of these is the Application Participant, which participates in discovery so that the Handler can route messages efficiently.  The Relay Participant operates in a different DDS domain for the purpose of gossiping with other RtpsRelay instances for horizontal scalability.  See [this video](https://www.youtube.com/watch?v=-WzgzKrWmKA) for the details.

The vertical ports are UDP and not TCP.  Adding TCP is on the roadmap.

It may be possible to configure the Application Participant to discover participants running in the same network/VM/etc.  However, this would cut them out of the Handler.  Consequently, they could talk to each other and the Application Participant, but not any other client participant who is using the vertical handlers.

I would like better to understand your concern with `SpdpRtpsRelayAddress` and friends.  Other users of the RtpsRelay do not statically configure `SpdpRtpsRelayAddress` and friends.  Instead, they
1. Create a JSON file containing the configuration that a client would need to use a particular instance of an RtpsRelay. 
2. Turn on the embedded meta-discovery server to serve the JSON file.
3. Put an HTTP load balancer in front of all of the meta-discovery servers.
4. On start up, the client participant accesses the HTTP load balancer which returns the JSON for one of RtpsRelay instances.
5. The client participant parses the JSON and dynamically sets `SpdpRtpsRelayAddress` and friends.
6. The client monitors the BITs to see if it is still connected to the RtpsRelay and goes back to step 4 if not.

We have not gone so far as to canonicalize this, but it seems like we will get there eventually.

Note that client participants will prioritize local discovery and communication if possible.  That is, if A and B can communicate locally but C cannot and all three are using the RtpsRelay, then A and B will exchange data using local multicast/unicast while communication with C will go through the RtpsRelay.

**Comment by smnrgrs** [comment]
Thanks for your reply as always. To explain my situation:

We have run an RTPSRelay in the cloud, and it allows our LAN participants to successfully connect/interoperate.

We then moved on to running participants and relays on a cloud server, and our LAN participants connected happily and interoperate.

I'm now looking at using the Cloud Multicast Repeater to enable multiple cloud server instances (and regions) to run participants - this works fine but our LAN participants, connecting via a cloud RTPSRelay, only see services which are local to the RTPSRelay and which connect directly to the RTPSRelay.

For my use case what I really need is an RTPSRelay whose application participant is capable of local participant discovery, as well as direct UDP connections. Then we could have multiple cloud server instances with multiple relays, LAN participants would connect to relays in the closest cloud region, and multicast repeaters would bridge the cloud regions together.

Hopefully that explains where the original question was coming from.

**Comment by jrw972** [comment, marked as answer]
I think I'm starting to understand.  You started with an RtpsRelay with LAN Participants.
```
                +-----------+
                | RtpsRelay |
                +-----------+
                  ^       ^
                  |       |
                  |       |
  +-----------------+   +-----------------+
  | LAN Participant |   | LAN Participant |
  +-----------------+   +-----------------+
```

You then expanded to LAN and cloud participants
```
              +-----------+
              | RtpsRelay |
              +-----------+
                ^       ^
                |       |
                |       |
+------------------+  +---------------------+
| LAN Participants |  | Cloud Participants  |
+------------------+  +---------------------+

```

And now you want to deploy multiple RtpsRelays and Cloud Participants in different regions and possibly different clouds.
```
              +-----------+                                  +-----------+
              | RtpsRelay |<-------------------------------->| RtpsRelay |
              +-----------+                                  +-----------+
                ^       ^                                      ^       ^
                |       |                                      |       |
                |       |                                      |       |
+------------------+  +---------------------+  +------------------+  +---------------------+
| LAN Participants |  | Cloud Participants  |  | LAN Participants |  | Cloud Participants  |
+------------------+  +---------------------+  +------------------+  +---------------------+
```

> For my use case what I really need is an RTPSRelay whose application participant is capable of local participant discovery, as well as direct UDP connections. Then we could have multiple cloud server instances with multiple relays, LAN participants would connect to relays in the closest cloud region, and multicast repeaters would bridge the cloud regions together.

By design, Application Participants only discover client participants that use the (Vertical) Handler.  Going through the Vertical Handler means that the relay can determine the IP and port for a particular client.  The Application Participant engages in discovery so that it can learn about the writers/readers and their partitions.  It then uses this information to optimize routing.  Allowing the Application Participant to discover a client participant that is not using its associated Vertical Handler will not help to establish connectivity.

However, connecting the Relay Participants and the horizontal handlers would provide the connectivity you seek.  I described this in https://github.com/OpenDDS/OpenDDS/discussions/4324.  However, here are some additional thoughts.

The horizontal handlers can and should be converted to DDS topics.  This would make securing them much easier.

Assuming this is done, here are some approaches that don't introduce additional network overlay technologies like VPN.

First, suppose there are two-tiers of RtpsRelay in your system.  The top-tier consists of a small RtpsRelay instance or cluster.  This instance or cluster exists to ensure connectivity between the Relay Participants in the second tier which could be distributed over multiple regions, clouds, etc.

Second, one could use the multicast repeater to link all of the RtpsRelay instances in the system.  Currently, there would be three repeaters per RtpsRelay instance but it would probably be straightforward to reduce this to one.  The downside of this approach is that you would have to open up firewall ports so that the multicast repeaters could communicate and you would need to ensure the multicast repeaters are fed with the IP:ports of all the other repeaters.

Third, one could extend the RtpsRelay to behave more like a multicast repeater.  That is, any data that it receives on a vertical port would be sent to all local clients and a configurable but fixed set of addresses (the horizontal ports of other relay instances) and anything received on a horizontal port is sent to all clients on the vertical ports.  This mode might be used in the first tier of the first approach. 

To summarize, the key functionality to making your scenario work is establishing connectivity between the Relay Participants and horizontal ports of the RtpsRelay.
